### ISSUE TITLE
Cannot load plugins

### ISSUE THREAD

**Issue by sleepyeye** [mosra/magnum#320]
Hello,

I'm trying to load bunch of images and meshes using various importer.

For example, I tried to codes in below to import image files.
```cpp
  PluginManager::Manager<Trade::AbstractImporter> manager;
  Containers::Pointer<Trade::AbstractImporter> importer =
      manager.loadAndInstantiate("AnyImageImporter");
  if (!importer || !importer->openFile("image.png"))
    Fatal{} << "Can't open image.png with AnyImageImporter";
```

However the code above spit out run time error message like this

```
 PluginManager::Manager::Manager(): none of the plugin search paths in {magnum/importers, /usr/local/lib/magnum/importers} exists and pluginDirectory was not set, skipping plugin discovery
PluginManager::Manager::load(): plugin AnyImageImporter is not static and was not found in 
Can't open image.png with AnyImageImporter
```

Since there is no `magnum/importers` folder around my executable in build directory, the error message seems normal. Also, `extern/magnum/importers` folder does not exist.

I followed the viewer example for writing cmake and my importer source codes.
The codes are almost identical except project structure.

My project structure looks like this.

```
.
├── CMakeFiles
├── build
│   ├── CMakeFiles
│   ├── extern
│   │   ├── corrade
│   │   ├── eigen
│   │   ├── glfw
│   │   └── magnum // does not contain importers folder
│   ├── src
│   └──  executable
├── extern
│   ├── corrade
│   ├── eigen
│   ├── glfw
│   └── magnum 
├── modules
└── src
    └── Myproject
```

1. What am I missing now? Why the `magnum/importer` folder does not generated in my build folder?
Is there any other cmake option for these?
2. In case of I should change the plugin directory. How can I override the default plugin search path?

**Comment by sleepyeye** [comment]
For question number 2, I found the answer
https://doc.magnum.graphics/corrade/classCorrade_1_1PluginManager_1_1Manager.html#PluginManager-Manager-paths


But still cannot figure out why my project cannot produce `magnum/importer` folder around my executable.

**Comment by mosra** [comment]
Hi!

the expected directory structure for plugins is created once you install the project to some location (`make install` and equivalent). CMake by default installs to `/usr/local`, so that's why the message says it's looking into `/usr/local/lib/magnum/importers`. Having the Magnum & Corrade dependencies installed makes sense if you have them built separately (or for example installed via your distribution/OS package manager). However, since you're using the CMake subproject approach here, installing everything in order to run it makes less sense. There are the following options:

- Build & install the dependencies separately instead  -- if you are on ArchLinux, Debian/Ubuntu, Gentoo or macOS, there's already a ready-to-use package for your system (and Eigen + GLFW as well) and then the plugins will appear in `/usr/local/lib` or `/usr/lib`. This might or might not be an option for you, depending on what you plan to do.
- Load the plugin by passing the whole path to wherre it got built to `loadAndInstantiate()`, for example `loadAndInstantiate("/home/you/code/yourproject/build/extern/magnum/src/MagnumPlugins/AnyImageImporter/AnyImageImporter.so")`. That's more of a hack, it hardcodes the CMake build directory structure (even though you can use e.g. [Directory::executableLocation()](https://doc.magnum.graphics/corrade/namespaceCorrade_1_1Utility_1_1Directory.html#a4e3dbc86c7b8d0c3c39d9802a9947b35) to parametrize at least the project location) and you'd also lose all advantages of the `AnyImageImporter` plugin auto-discovery -- for example, to load PNGs, you'll need to load the `PngImporter.so` manually as well, it won't be able to find it on its own this way.
- Switch to [static plugins workflow](https://doc.magnum.graphics/magnum/plugins.html#plugins-static). The C++ side will not change, but you'll need to explicitly link the needed plugins via CMake. I think this is the most feasible solution if you plan to stay with the CMake subprojects approach.

Note: there's a [Magnum Plugins](https://github.com/mosra/magnum-plugins) repository, containing the actual full-featured set of plugins. What's contained in the core Magnum repository is just a barebones set supporting not much more than OBJ and TGA.

Also, seeing you're using Eigen, this work-in-progress might be of interest for you: https://github.com/mosra/magnum-integration/pull/40

Hope this helps! :)

**Comment by sleepyeye** [comment]
I just took a look magnum's main `CMakeLists.txt` file.
I saw quite a lot of  `MAGNUM_PLUGINS_*_DIR` variables.
Most of them depends on `*_INSTALL_DIR` variables. 

Instead of installing Magnum in specific directory, I used it as subproject (by using git submodule, cmake add_subdirectory). 
I'm building my projects like this
```
cd build
cmake ..
make
```

This means that there is no chance to happen `make install` command. Thus, plugins will not be built/installed.
So lack of `magnum/importers` library is normal behavior if I only trigger `make` command.
Am I right?


If it is right, is there any other option to avoid it?
Should I use static importers instead dynamic one?

**Comment by mosra** [comment]
Haha, we replied at the same time :) Tell me if anything from the above list helps.

**Comment by sleepyeye** [comment]
@mosra 
Thank you for your kind answer again.
TL;DR I decided to use static plugin workflow with Magnum plugin libs

1.
Even though I'm using Ubuntu and I can install all my external libs in `/usr/local`.
I put all the ingredients in subproject so that it can be built and run in different environments.
I decided this because my project will be used in multiple environment and portability is important for end-user.
But thank you for your kind advice.

2. 
I decided to use static plugin workflow with Magnum plugin libs. That looks like the best option for me. But I found some weird things.
There is no `/home/you/code/yourproject/build/extern/magnum/src/MagnumPlugins/AnyImageImporter/AnyImageImporter.so` file in my build directory.
Is this normal? Or am I miss something? (I *DIDNT* import the Magnum plugin library yet!)


3. 
Yes. I am very interested in Eigen integration with Magnum. Thanks a lot.
I didn't know you are already working on this but if you were not, I would suggest the option for it.
I read some commits but it seems like commits only contains type conversion between `Eigen` and `Magnum::Math` just like `Magnum::GlmIntegration`.

Currently it is okay to me now. But I just want to ask that do you have any plan that use `eigen` as backend option for `Magnum::Math`? (means replacing entire `Magnum::Math` operation with `eigen` and use `Magnum::Math` as interface or wrapper for `eigen`)

**Comment by mosra** [comment]
1. Okay, I understand. It doesn't necessarily *have* to be in `/usr/local`, building & installing next to where the project is cloned can work as well. But I agree that the subproject approach is more portable and easier to use :)
2. If you have it enabled, then there should be *some* `AnyImageImporter.so` file somewhere in the build directory. I wrote the path off the top of my head, so I might have made a mistake there :)
3. No, there's definitely no plan to replace `Magnum::Math` with Eigen (or any other library, for that matter), only allowing them to coexist easier. The use cases of Eigen and `Magnum::Math` are vastly different (magnum math is graphics-oriented, while Eigen is more scientific), moreover Eigen is very slow to compile compared to Magnum and the usability and documentation and feature set for graphics applications is also not at a level I would like. There's a lot more to say about that -- in the near future there will be a post [on the blog](https://blog.magnum.graphics) comparing magnum math to other implementations, and showing various reasons why I chose to not reuse any existing library.

**Comment by sleepyeye** [comment]
@mosra 
Ohhhh yeah you are right.  Doing scientific computation using `Magnum::Math` doesn't sound right. 
That is totally out of scope. My bad.

Also I'll keep my eye on the blog. Since the topic sounds really interesting.

Thank you!
